### PR TITLE
Rev Alpine WithNode from 3.9 to 3.13

### LINKED ISSUE #56672: Alpine builds use alpine-3.9 docker image that is EOL
Alpine builds run on alpine-3.9-WithNode-20210714125437-9b5bbc2 image . Alpine 3.9 is EOL (https://alpinelinux.org/releases/). We should update the builds to run on supported Alpine version.

### PR DESCRIPTION
**Pull request by am11**
Also,
* update helix queues from 3.12 to 3.13
* add 3.14 helix queues
* update performance job to install cargo from main branch

Fixes #56672

### CODE REVIEW

**Comment by am11** [comment]
cc @jkotas, @janvorli we can also update helix queues to test the latest Alpine Linux version, but I haven't changed that as part of this PR.

**Comment by jkotas** [comment]
According to https://github.com/dotnet/core/blob/main/release-notes/6.0/supported-os.md#linux, we want to support Alpine 3.13+ for .NET 6. Are binaries built on Alpine 3.14 going to run on Alpine 3.13?

**Comment by am11** [comment]
> Are binaries built on Alpine 3.14 going to run on Alpine 3.13?

They should work (as time64 changes were included in 3.13). I will test it.

**Comment by am11** [comment]
Locally I was able to build the main branch without any changes:

```sh
docker run -v $(pwd):/runtime \
  -it mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-WithNode-20210812132848-d90babf \
  /runtime/src/coreclr/build-runtime.sh checked x64
```

but on CI it was failing to find python while it is clearly there, but when i worked around it, it is now failing with `Failed to get PGO data package path`. There is nothing in logs pointing to clues what's wrong, so it might be something specific to pipeline YAML that needs updating?

**Comment by hoyosjs** [comment]
@am11 looks like the container doesn't have curl? (and maybe doing the volume mount masks this as you may have a `.dotnet` directory?

**Comment by am11** [comment]
Localhost:
```
% docker run -it mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-WithNode-20210812132848-d90babf /bin/sh
/ # curl --version
curl 7.78.0 (x86_64-alpine-linux-musl) libcurl/7.78.0 OpenSSL/1.1.1k zlib/1.2.11 brotli/1.0.9 nghttp2/1.43.0
Release-Date: 2021-07-21
Protocols: dict file ftp ftps gopher gophers http https imap imaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp 
Features: alt-svc AsynchDNS brotli HSTS HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM NTLM_WB SSL TLS-SRP UnixSockets
```

CI:
```
dotnet_install: Error: curl (recommended) or wget are required to download dotnet. Install missing prerequisite to proceed.
```

it is definitely not running in the same container.
@wfurt, @hoyosjs, do you have any clues what might be the cause?

**Comment by am11** [comment]
> @am11 looks like the container doesn't have curl? (and maybe doing the volume mount masks this as you may have a `.dotnet` directory?

@hoyosjs, I tried  it with clean git repo. Even without mounting, `git clone <runtime url>` followed by `build-runtime.sh release x64` succeeds in that container.

**Comment by hoyosjs** [comment]
I tried with copied IDs from the build logs and I was still able to get the SDK.

**Comment by am11** [comment]
Yes, the docker container `mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-WithNode-20210812132848-d90babf` is the one failing CI leg is using which I tested earlier. It has all the required stuff and capable of building runtime. Something else in AzDo is apparently going wrong. Nothing jumping out at me from logs of `Initialize containers` step, all looks normal until `Build CoreCLR Runtime` step, which seems to be using a different environment.

**Comment by ViktorHofer** [comment]
@am11 please let us know if we can help anyhow.

**Comment by am11** [comment]
@ViktorHofer, it is still a mystery to me as CI error is not reproducible locally. e.g. this leg https://dev.azure.com/dnceng/public/_build/results?buildId=1291489&view=logs&jobId=0e64859c-a870-5a74-6b6e-333fc1003298&j=0e64859c-a870-5a74-6b6e-333fc1003298&t=5da9eef6-c013-592b-ccae-2e46c52786d0 is apparently using container `mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-WithNode-20210812132848-d90babf` and failing on build step due to missing `curl`. Although the said container has `curl`, `python` along with reset of the dependencies.

I just (temporarily) deleted ` > /dev/null 2>&1 ` to reveal the missing curl error from https://github.com/dotnet/runtime/blob/6acb72bb66c3b68672a0653b35a193b39677c33b/src/coreclr/build-runtime.sh#L55-L58

If we test it locally with same command used by this failing CI leg:

```
docker run \
  -it mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-WithNode-20210812132848-d90babf \
  sh -c 'git clone https://github.com/dotnet/runtime --single-branch --depth 1 /runtime &&
  /runtime/src/coreclr/build-runtime.sh checked x64'
```
it builds coreclr successfully. Do you see anything unusual from the logs that could be causing `Build CoreCLR Runtime` step to use different environment than the container?

**Comment by ViktorHofer** [comment]
I'll take a look. Let me also cc @safern and @jkoritzinsky

**Comment by ViktorHofer** [comment]
@hoyosjs @jkoritzinsky any other ideas? I'm currently a bit short on time, maybe you have some additional ideas?

**Comment by ViktorHofer** [comment]
@MattGal do you think you can help here?

**Comment by MattGal** [comment]
> @MattGal do you think you can help here?

No guarantees but sure, I'll take a look

**Comment by MattGal** [comment]
@ViktorHofer I can't reproduce this running main build inside a `mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-WithNode-20210812132848-d90babf` container, it may be a transient error?

**Comment by hoyosjs** [comment]
Sigh. Looks like transient - cache or something. Kicked off again and the containers initialized just fine...

**Comment by MattGal** [comment]
Super weird but does seem to be noticed and known externally: https://github.com/dotnet/install-scripts/issues/206.

I'm running the same image you used to run the same CI build from a Windows box.  I wonder if you started from an Ubuntu 18.04/20.04 if you could get a repro.

**Comment by hoyosjs** [comment]
I couldn't repro on a physical Ubuntu box with that container on top.

**Comment by am11** [comment]
Funny stuff from logs:

```sh
command -v python3 => /usr/bin/python3
# (next line from our script which makes sure it's really there)
command -v /usr/bin/python3 => {false}
# something that i added for testing
/usr/bin/python3 -c "print('/usr/bin/python3 exists')" => /usr/bin/python3 exists
```

so `command -v` with command name is working, with relative and absolute path the executable is working fine too, but with absolute path command -v fails..

commenting out `exit 1` for python lead to the next issue `dotnet-install.sh` not finding curl and wget, while curl is clearly installed. In that instance, `hash curl` returns false in the CI.

Locally it just works,  neither of those issues (`command -v /usr/bin/python3`, `hash curl`) exist.

---

Also from logs:

```bash
Starting: Initialize containers
/usr/bin/docker version --format '{{.Server.APIVersion}}'
'1.40'
Docker daemon API version: '1.40'
/usr/bin/docker version --format '{{.Client.APIVersion}}'
'1.40'
Docker client API version: '1.40'
```
---
Conclusion (kinda):
I'm convinced that 3.14 container itself has everything that we need to build runtime, but the base container (`library/alpine:3.14`) was built with newest docker builder toolchain (buildx et al.) which seem to have a breaking change. For the new containers, we would need AzDO to support newer docker that contains this runc patch: https://github.com/opencontainers/runc/pull/2750.

For now, we can stick to Alpine Linux 3.13 (that has EOL in Nov. 2022). We can try Alpine 3.14 later, when we start to see docker higher than 1.40 in logs.

**Comment by ViktorHofer** [comment]
@am11 we are still using some old alpine3.12 queues for helix testing. Do you want to update these as part of this PR as well? Fine either way.

- https://github.com/dotnet/runtime/blob/f41b259d779466068d5f7b1a16608e74e26a5e46/eng/pipelines/libraries/helix-queues-setup.yml#L45-L52
- https://github.com/dotnet/runtime/blob/f41b259d779466068d5f7b1a16608e74e26a5e46/eng/pipelines/coreclr/templates/helix-queues-setup.yml#L66-L85

**Comment by am11** [comment]
@ViktorHofer, for 3.14, all platforms were added (in dotnet/dotnet-buildtools-prereqs-docker@d90babf) which were available on 3.9. Unfortunately, the preexisting 3.13 didn't have helix support for all architectures (only armv7 is available; missing arm64 and amd64). I can put together 3.13-helix PR in prereqs repo separately (codeflow will take a while; a day or two).

**Comment by ViktorHofer** [comment]
> @ViktorHofer, for 3.14, all platforms were added (in dotnet/dotnet-buildtools-prereqs-docker@d90babf) which were available on 3.9. Unfortunately, the preexisting 3.13 didn't have helix support for all architectures (only armv7 is available; missing arm64 and amd64). I can put together 3.13-helix PR in prereqs repo separately (codeflow will take a while; a day or two).

We would like to test against 3.13 on PRs and 3.14 on rolling builds (isFullMatrix switch in the libraries helix queues template). 

> I can put together 3.13-helix PR in prereqs repo separately (codeflow will take a while; a day or two).

Wow that would be fantastic. Really appreciate your help here.

cc @aik-jahoda who owns updating our helix queues for main and .NET 6 which is (not yet up to date) tracked in: https://github.com/dotnet/runtime/issues/57947

**Comment by am11** [comment]
@ViktorHofer, updated test jobs with new alpine 3.13 helix images. Could you please run outerloop pipeline?

CMIIW, but in outerloop jobs, we use the same container to build / package projects under `src/tests` before sending to helix, right? 3.14 helix images are also available, but because it will try to use that container on AzDO first for packaging, it will fail.

**Comment by ViktorHofer** [comment]
> CMIIW, but in outerloop jobs, we use the same container to build / package projects under src/tests before sending to helix, right? 3.14 helix images are also available, but because it will try to use that container on AzDO first for packaging, it will fail.

Tbh don't know about how src/tests build in CI. If that's the case, then 3.13 is just fine for these tests.

**Comment by am11** [comment]
@ViktorHofer, I added 3.13 as well as 3.14 helix queues in libraries and coreclr for testing. The libraries tests have passed on both queues. 👍

Could you please re-run outerloop so we will find out if it works with 3.14 helix containers?

ps - it is a bit odd that while libraries has three musl test legs in `runtime` pipeline, we don't have a single coreclr alpine/musl test leg in `runtime` or `runtime-staging` pipelines that are triggered with PR. All musl test legs for coreclr are in outerloop. Maybe we should move at least one of musl test legs to `runtime` pipeline? 🤔

**Comment by ViktorHofer** [comment]
> ps - it is a bit odd that while libraries has three musl test legs in runtime pipeline, we don't have a single coreclr alpine/musl test leg in runtime or runtime-staging pipelines that are triggered with PR. All musl test legs for coreclr are in outerloop. Maybe we should move at least one of musl test legs to runtime pipeline? 🤔

That's a question for the coreclr infra crew. cc @trylek @jkoritzinsky @hoyosjs. Might make sense to open a separate issue for discussing this.

**Comment by jkotas** [comment]
>  we don't have a single coreclr alpine/musl test leg

coreclr test legs are relatively expensive with the current test setup. I am not sure whether it is intentional or not, but it does not sound wrong to me that there are no apline/musl legs in the default PR CI matrix.

**Comment by am11** [comment]
Blocked on https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/501.

**Comment by am11** [comment]
musl legs are running normal. Lets see if outerloop legs agree with the new setup. 🤞

**Comment by am11** [comment]
Alpine side of matrix is green 🌴 including 3.14 helix. WithNode is still on 3.13 due to AzDO running old docker. We can retry next year when 3.13 will reach its EOL (in Nov. 2022).

rest  of the failures are unrelated to PR changes (failing on main as well).

**Comment by am11** [comment]
Any other comments, good to merge?

**Comment by ViktorHofer** [comment]
@wfurt can you please also take a look at the changes?

**Comment by hoyosjs** [comment]
We are adding more pressure to the Linux Queues (double the libraries musl WI for PRs I think?). Do we want to just test one and let rolling test the rest?

**Comment by am11** [comment]
Right, it wasn't by mistake; I have followed @ViktorHofer's guidance from the other PR: https://github.com/dotnet/runtime/pull/57841#discussion_r693090482. i.e. it is easier to disable (basically comment out a line in yaml..) later once we make sure that the official matrix is supported.

**Comment by ViktorHofer** [comment]
@safern can you please take a look at this PR?

File: eng/pipelines/libraries/helix-queues-setup.yml
Comment (hoyosjs) [inline eng/pipelines/libraries/helix-queues-setup.yml:45]: Both branches are the same - we can probably drop the conditions. Even better - only add 3.14 on `isFullMatrix=true` as it's unlikely a PR will see benefit from testing both.

File: eng/pipelines/coreclr/templates/helix-queues-setup.yml
Comment (hoyosjs) [inline eng/pipelines/coreclr/templates/helix-queues-setup.yml:69]: Now that we confirmed it worked for both, can you please comment one (probably 3.14)? CoreCLR testing is quite expensive.

**Review by hoyosjs — APPROVED** [review]
Other than the couple comment, looks great. Thanks for pushing on with this! That was a pretty gnarly set of looping.

File: eng/pipelines/coreclr/templates/helix-queues-setup.yml
Comment (am11) [reply eng/pipelines/coreclr/templates/helix-queues-setup.yml:69]: Given it runs in outerloop tests in parallel on Helix (when PRs are merged and no one is 'waiting'), do you think 3.13-only is good enough? Empirically speaking, these legs typically complete much faster (<= glibc distros' legs). Most time is spent on simulator and mobile platform jobs (which offsets everything else in terms of timing).

Comment (hoyosjs) [reply eng/pipelines/coreclr/templates/helix-queues-setup.yml:69]: I defer to @jkotas, he's investigated the impact of our testing on resources more heavily lately. My main qualm is the Helix queue this is running on, the ubuntu.1604.amd64 variants, has 200 machines shared for every dotnet repo and it's got pretty high demand. Even if running fast, they can only take so much work in parallel and everyone wants them. The 5.000.000+ tests per CI are already stressing that - add more to that per run and it will creep on us pretty badly; until we have better retries and a cleaner outerloop run, adding to outerloop is just adding to our pressure and to a long list of "yeah, things are broken. File an issue,"  so I've been generally skeptic on the value we get vs how expensive it is.

File: eng/pipelines/libraries/helix-queues-setup.yml
Comment (am11) [reply eng/pipelines/libraries/helix-queues-setup.yml]: Removed 3.13 from isFullMatrix=true.

---
A side note: in some places, we are using this syntax as here. or more generally:
```yaml
- ${{ if eq(someParm, false) }}:
  - foo
- ${{ if eq(someParm, true) }}:
  - bar
```
in others, it is different:

```yaml
- ${{ if eq(someParm, true) }}:
  - foo
- ${{ if ne(someParm, true) }}:
  - bar
```
unless the latter is to cover `true` or  `<everything else>`, it would be nice to define a convention as YAML + this custom expression syntax ain't easy to digest. 😅

File: eng/pipelines/libraries/helix-queues-setup.yml
Comment (hoyosjs) [inline eng/pipelines/libraries/helix-queues-setup.yml:48]: Sorry, I meant more along the lines of:
```yml
    - ${{ if eq(parameters.platform, 'Linux_musl_x64') }}:
      - (Alpine.313.Amd64.Open)ubuntu.1604.amd64.open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.13-helix-amd64-20210910135845-8a6f4f3
      - ${{ if eq(parameters.jobParameters.isFullMatrix, true) }}:
        - (Alpine.314.Amd64.Open)ubuntu.1604.amd64.open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-amd64-20210910135833-1848e19
```
If we believe alpine 3.14 brings enough value.

File: eng/pipelines/libraries/helix-queues-setup.yml
Comment (hoyosjs) [reply eng/pipelines/libraries/helix-queues-setup.yml]: The thing is in yaml things decay to strings relatively easily. It's safer to do eq true, and ne true.

Comment (am11) [reply eng/pipelines/libraries/helix-queues-setup.yml]: Yup, so latter one is better if values are different.

File: eng/pipelines/libraries/helix-queues-setup.yml
Comment (am11) [reply eng/pipelines/libraries/helix-queues-setup.yml:48]: > If we believe alpine 3.14 brings enough value.

I think the overarching idea is: build on oldest, test (at least) on latest (to maximize compatibility coverage). That goes for all distros, including Alpine, which is why I believe 3.14 is good to keep.

File: eng/pipelines/coreclr/templates/helix-queues-setup.yml
Comment (am11) [reply eng/pipelines/coreclr/templates/helix-queues-setup.yml]: Commented out 3.13 helix legs.

File: eng/pipelines/coreclr/templates/helix-queues-setup.yml
Comment (safern) [inline eng/pipelines/coreclr/templates/helix-queues-setup.yml:69]: Why are we commenting 313 ones instead of just deleting them?

File: eng/pipelines/libraries/helix-queues-setup.yml
Comment (safern) [inline eng/pipelines/libraries/helix-queues-setup.yml:46]: Do we want to keep running 313 on PRs? -- that means on PRs we will run 313 and 314? Maybe we should keep 313 as rolling only? 

cc: @jkotas @aik-jahoda @ViktorHofer

Comment (am11) [reply eng/pipelines/libraries/helix-queues-setup.yml:46]: Agreed. Should 3.13 vs. 3.14 (latest; released 3 months ago) follow the same logic which @aik-jahoda used for Ubuntu (https://github.com/dotnet/runtime/pull/59578)?

Comment (jkotas) [reply eng/pipelines/libraries/helix-queues-setup.yml:46]: We should only have one Alpine version on PRs.

Comment (safern) [reply eng/pipelines/libraries/helix-queues-setup.yml:46]: In my opinion running latest on PRs makes more sense.

Comment (am11) [reply eng/pipelines/libraries/helix-queues-setup.yml]: Updated with 314-only queues.

File: eng/pipelines/coreclr/templates/helix-queues-setup.yml
Comment (am11) [reply eng/pipelines/coreclr/templates/helix-queues-setup.yml]: Removed.

File: eng/pipelines/libraries/helix-queues-setup.yml
Comment (safern) [inline eng/pipelines/libraries/helix-queues-setup.yml:44]: can we test on 3.13 on fullMatrix == true? Meaning on rolling CI?

Comment (am11) [reply eng/pipelines/libraries/helix-queues-setup.yml:44]: Yup, didn't mean to delete on rolling ones 🤣

**Review by safern — APPROVED** [review]
Thanks 🎉

**Review by hoyosjs — APPROVED** [review]
Thanks a lot @am11 :shipit:

**Comment by safern** [comment]
It seems like you have a yaml syntax error: https://dev.azure.com/dnceng/public/_build/results?buildId=1384687&view=results

**Comment by aik-jahoda** [comment]
There is a JSON failure on Alpine.314.Arm64.Open. @safern do you think the failure is related to the change or we can merge this PR?

**Comment by danmoseley** [comment]
The JSON crash has a dump. We should look at that dump -- there is a how-to-debug-dump.md file there to help.

The regex issue is known and fixed already.

**Comment by am11** [comment]
> There is a JSON failure on Alpine.314.Arm64.Open

Isn't it https://github.com/dotnet/runtime/issues/57198?

**Comment by safern** [comment]
I think we should at least look at the dump to make sure if it is: #57198 before merging something that could potentially introduce a failure to CI.

**Comment by am11** [comment]
The dump link on that issue is expired and Eirik had no success with `clrstack` when link was active (https://github.com/dotnet/runtime/issues/57198#issuecomment-900433052). Not sure how I can determine it.

Maybe rerun the job to see if the issue persist?

**Comment by safern** [comment]
> Maybe rerun the job to see if the issue persist?

Sounds good to me.

**Comment by danmoseley** [comment]
Dump links from #57324 are still active, they're only hours old.

Console Logs

https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-57324-merge-9e9d1cbb6f224c0492/System.Text.Json.Tests/1/how-to-debug-dump.md?sv=2019-07-07&se=2021-10-19T14%3A08%3A15Z&sr=c&sp=rl&sig=wKHlTG%2F2YMVkeicpVl0W%2FlyaZ4RfBE74SXPTyvUs6b0%3D

https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-57324-merge-9e9d1cbb6f224c0492/System.Text.Json.Tests/1/how-to-debug-dump.md?sv=2019-07-07&se=2021-10-19T14%3A08%3A15Z&sr=c&sp=rl&sig=wKHlTG%2F2YMVkeicpVl0W%2FlyaZ4RfBE74SXPTyvUs6b0%3D

from
https://dev.azure.com/dnceng/public/_build/results?buildId=1385273&view=ms.vss-test-web.build-test-results-tab&runId=40500360&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab&resultId=185830

I believe it requires a certain access token to pull part of what you need down so @am11 you maybe can't.

**Comment by am11** [comment]
> Dump links from #57324 are still active, they're only hours old.

Yup, I can download dump from this PR but the one from https://github.com/dotnet/runtime/issues/57198 (which I would like to compare with) is expired.

**Comment by danmoseley** [comment]
Ah gotcha. Is there anything that can be gotten out of this dump?

**Comment by am11** [comment]
I am working on it. 🙂

**Comment by am11** [comment]
`clrstack` results are posted here: https://github.com/dotnet/runtime/issues/57198#issuecomment-930569859. I have the container saved, in case more info is needed.

**Comment by aik-jahoda** [comment]
> The regex issue is known and fixed already.

> `clrstack` results are posted here: [#57198 (comment)](https://github.com/dotnet/runtime/issues/57198#issuecomment-930569859). I have the container saved, in case more info is needed.

Sounds like both test failures are known and tracked/fixed already. Can we merge this PR?

**Pull request merged**
